Install files in parallel

Thomas1664 opened the pull request:
Depending on how many files to install, this change results in a >= 4x speed up. I only tested relatively small ports but I expect even higher speed ups for bigger ports.

1. Get file status in parallel
1. Write listfile (async)
1. Create all directories
1. Copy files in parallel

Thomas1664 commented on src/vcpkg/commands.install.cpp:
```diff
@@ -156,15 +193,20 @@ namespace vcpkg
+            });
 
+        std::move(dir_output.begin(), dir_output.end(), std::back_inserter(output));
+#ifdef _WIN32
+        std::sort(std::execution::par_unseq, output.begin(), output.end());
+#else
         std::sort(output.begin(), output.end());
+#endif
```
I'm open for suggestions on how to implement parallel sorting for all platforms

Thomas1664 commented on src/vcpkg/commands.install.cpp:
```diff
@@ -73,60 +75,99 @@ namespace vcpkg
-                case FileType::directory:
-                {
-                    fs.create_directory(target, ec);
-                    if (ec)
-                    {
-                        msg::println_error(msgInstallFailed, msg::path = target, msg::error_msg = ec.message());
-                    }
+                msg::println_error(msgInvalidFileType, msg::path = files[i]);
+                continue;
+            }
+            files_and_status.emplace_back(std::move(files[i]), std::move(symlink_statuses[i]));
+        }
 
-                    // Trailing backslash for directories
-                    this_output.push_back('/');
-                    output.push_back(std::move(this_output));
-                    break;
-                }
-                case FileType::regular:
+        std::vector<std::string> output(files_and_status.size());
+
+        parallel_transform(
+            files_and_status.begin(), files_and_status.size(), output.begin(), [&](auto&& file_and_status) {
+                auto&& [file, maybe_status] = file_and_status;
+
+                const auto status = *maybe_status.get();
+                std::error_code ec;
+
+                const auto suffix = file.generic_u8string().substr(prefix_length + 1);
+                const auto target = destination / suffix;
+
+                bool use_hard_link = true;
+                auto this_output = Strings::concat(destination_subdirectory, "/", suffix);
+                if (is_regular_file(status))
                 {
                     if (fs.exists(target, IgnoreErrors{}))
                     {
                         msg::println_warning(msgOverwritingFile, msg::path = target);
-                        fs.remove_all(target, IgnoreErrors{});
+                        fs.remove(target, IgnoreErrors{});
```
There's no difference for regular files between `remove_all` and `remove` (provided that vcpkg's implementation is identical to the STL): https://en.cppreference.com/w/cpp/filesystem/remove

Thomas1664 commented:
@BillyONeal Unit tests on OSX failed [here](https://github.com/microsoft/vcpkg-tool/actions/runs/6693816217/job/18185796764). It seems to work again but I can consistently reproduce the failure on my local machine: The test case `captures-output` fails with `error: calling CreateProcessW failed with 2 (unknown error)`. I'm not sure if this is the same test that fails in CI. I can confirm that this is most likely not caused by changes in this PR because the changed code inside `commands.install.cpp` is never hit.

autoantwort commented:
Btw you can build the format target locally to format all code.

BillyONeal commented:
Sorry!

BillyONeal commented on src/vcpkg/commands.install.cpp:
```diff
@@ -40,130 +44,202 @@ namespace vcpkg
+
+        parallel_transform(files, output.begin(), [&](auto&& file) -> PathAndType {
+            std::error_code ec;
+            auto status = fs.symlink_status(file, ec);
```
I strongly suspect the only thing you're gaining from parallelism here is parallelizing this `symlink_status` and this would be much clearer if you only parallelized that rather than trying to make invalid states that need to be chopped off etc.

If you parallelize just this part you could use plain `Util::erase_remove_if` for the invalid ones. It would also have the advantage of not trying to parallelize `println_warning`.

Thomas1664 replied:
If I want to get rid of invalid states I'd have to use `parallel_for_each` but this would require a lock on when inserting things into the destination vector. The only thing I can move out of this is the validation of the file type.

BillyONeal commented on src/vcpkg/commands.install.cpp:
```diff
@@ -40,130 +44,202 @@ namespace vcpkg
+            auto status = fs.symlink_status(file, ec);
+            if (ec)
+            {
+                msg::println_warning(format_filesystem_call_error(ec, "symlink_status", {file}));
```
`println_warning` is not thread safe.

BillyONeal commented on src/vcpkg/commands.install.cpp:
```diff
@@ -40,130 +44,202 @@ namespace vcpkg
+
+            if (fs.exists(target, IgnoreErrors{}))
+            {
+                msg::println_warning(msgOverwritingFile, msg::path = target);
```
`println_warning` is not thread safe.

BillyONeal commented on src/vcpkg/commands.install.cpp:
```diff
@@ -40,130 +44,202 @@ namespace vcpkg
                     if (ec)
                     {
-                        msg::println_error(msgInstallFailed, msg::path = target, msg::error_msg = ec.message());
+                        Debug::println("Install from packages to installed: Fallback to copy "
```
`Debug::println` is not thread safe.

Thomas1664 replied:
There is other parallelized code where this wasn't a problem: in `cmd_execute_and_capture_output_parallel` we call `cmd_execute_and_stream_data` which calls `Debug::print()` (assuming you'd consider `Debug::print` also not thread safe).

BillyONeal commented on src/vcpkg/commands.install.cpp:
```diff
@@ -40,130 +44,202 @@ namespace vcpkg
 
             bool use_hard_link = true;
-            auto this_output = Strings::concat(destination_subdirectory, "/", suffix);
-            switch (status)
+            std::error_code ec;
```
Please move these into the scope where they are used.

BillyONeal reviewed (changes requested):
I'm in general happy with attempts to parallelize things but overall I think you need to be much more explicit in proving that everything you call in a parallel context is actually safe to do. Given that the first run of e2e tests experienced a failure and I found races in a cursory review makes me fear that unknown races remain without such a proof.

This means if you want to proceed here I think you need to show strong thread safety requirements of every API you're calling rather than taking large blocks of existing code and wrapping them in parallel constructs.

Thomas1664 commented on src/vcpkg/commands.install.cpp:
```diff
@@ -40,130 +44,202 @@ namespace vcpkg
+            }
+        }
+
+        fs.create_directories(listfile.parent_path(), VCPKG_LINE_INFO);
```
Why? The list file is not in the same folder where the installed files are. And the "destination directory" is created beforehand. In theory, the only race condition would be for creating `/installed` but this already exists (line 163). The listfiles go in `/installed/vcpkg/info` and the installed files will go in `installed/triplet`.

Thomas1664 commented:
> I'm in general happy with attempts to parallelize things but overall I think you need to be much more explicit in proving that everything you call in a parallel context is actually safe to do. Given that the first run of e2e tests experienced a failure and I found races in a cursory review makes me fear that unknown races remain without such a proof.

I think the main issue with parallelizing here is printing and the combination of print and exit. While printing alone can be guarded by mutexes, this will result in deadlocks in combination with exit. Therefore, we need to make printing inherently thread safe.

Thomas1664 commented on src/vcpkg/commands.install.cpp:
```diff
@@ -40,130 +44,202 @@ namespace vcpkg
+
+        bool operator<(const PathAndType& other) const noexcept
+        {
+            if (!path.empty() && other.path.empty())
```
Note that the condition needs to be inverted because we want non-empty paths first.

Thomas1664 commented:
@BillyONeal now that #1323 landed, what needs to be done for this to land?

I checked each function for thread safety. The following functions are not thread safe because they print to the console:
- `fs.create_directories`
- `fs.write_lines`
- Obvious calls to print functions

I'm not entirely sure how to use `DiagnosticContext` here...

Thomas1664 commented:
@BillyONeal I addressed your concerns regarding conflicting directory creations and interleaved terminal output. I changed the code to write the list file from the main thread and I use a mutex to protect terminal output.
